Use dotted version vectors on common tests

lrascao opened the pull request:
Instead of keeping a large ets table of all messages seen, keep a dvv per stored object and use that instead in the plumtree behaviour

lrascao committed:
Set debug log level when running tests

lrascao committed:
Improve debug

lrascao committed:
Use dotted version vectors for causality tracking when running tests

cmeiklejohn commented on .travis.yml:
```diff
@@ -1,16 +1,28 @@
-  - make dialyzer
-  - make lint
+- ./rebar3 ct --suite test/plumtree_SUITE --group hyparview --case broadcast_high_client_test
+# - make xref
```
Why did you comment these out?

lrascao replied:
ah sorry, i should have mentioned it, this is to try and catch logs for this failed test, doesn't happen always

lrascao replied:
all debug commits are prefixed with `!drop` in the message

cmeiklejohn replied:
I don't understand?

cmeiklejohn replied:
So, should I be reviewing this for a merge or should I wait until the ```!drop``` commits are removed?

lrascao replied:
Waiting until i've found the root cause of the failed tests is maybe better, i'll ping again when that happens, meanwhile i'll mark as the PR as a WIP

cmeiklejohn commented:
Great, that helps.  Thanks.

lrascao committed:
Ensure success of forward message operation

lrascao committed:
Allow message id override on merge

When performing a broadcast merge it might be the
case that the application wants to return a new message
id that results from the merge and have that broadcast instead
of the originally received message id. Support this by handling
an extra return tuple from the callback.

lrascao commented:
@cmeiklejohn this is now ready for review

cmeiklejohn commented on src/plumtree_broadcast.erl:
```diff
@@ -451,12 +461,14 @@ send_lazy(#state{outstanding=Outstanding}) ->
 send_lazy(MessageId, Mod, Round, Root, Peer) ->
-    plumtree_util:log(debug, "sending lazy push ~p",
-               [{i_have, MessageId, Mod, Round, Root, myself()}]),
+    % plumtree_util:log(debug, "sending lazy push ~p",
```
Can we remove commented code?

lrascao replied:
done, fixup `694ce21` was pushed to that effect

cmeiklejohn commented on src/plumtree_broadcast.erl:
```diff
@@ -374,12 +375,17 @@ code_change(_OldVsn, State, _Extra) ->
     _ = send({prune, Root, myself()}, Mod, From),
     State1;
+handle_broadcast({true, MessageId}, _OldMessageId, Message, Mod, Round, Root, From, State) ->
+    handle_broadcast(true, MessageId, Message, Mod, Round, Root, From, State);
```
What is this new case for?

lrascao replied:
This is to overcome an issue i came across with. While performing a broadcast storm (ie. all members generating broadcasts) i would occasionally end up in a situation where one member would be rejecting `graft` requests as the context provided was stale (ie. another subsequent write had subsumed it). 
node A would generate a new write that gets eager pushed to node B, node B would then lazy push it to node C, node C doesn't have the key so it requests B to graft it but it is providing the context that originated from A, from the point of view of B  this context is stale, this way B never replies to C with the missing key.
This extra return tuple to `merge`  allows node B to return the new context after the merge and that's the one that gets lazy pushed to C.

cmeiklejohn replied:
Oh, that's neat.  Can you document this in the code with a comment to explain this very case?

lrascao replied:
done

cmeiklejohn commented on src/plumtree_broadcast.erl:
```diff
@@ -645,7 +657,7 @@ send(Msg, Mod, P) ->
     PeerServiceManager = PeerService:manager(),
     instrument_transmission(Msg, Mod),
-    PeerServiceManager:forward_message(P, ?SERVER, Msg).
+    ok = PeerServiceManager:forward_message(P, ?SERVER, Msg).
```
If this fails, will it crash?  Is that right?

lrascao replied:
It will, the idea is to not silently forward messages to failed peers

cmeiklejohn replied:
But, won't this crash the caller?  I'm wondering if this should return error instead?

lrascao replied:
it doesn't crash the caller but it does crash the `plumtree_broadcast` worker process, returning error here doesn't really help since it is being ignored on all invocations

lrascao replied:
another alternative is reporting an error so it has some visibility, thoughts?

cmeiklejohn reviewed (changes requested):
Can you explain why the DVV addition is valuable/necessary?

lrascao commented:
The DVV approach for tests is an alternative for the previous ETS based one which i'm not at all comfortable about going to production with (ie. a grow only table that keeps track of seen messages), most of the ideas/code were picked up from riak core and applied here.

lrascao commented:
there is one pending fixup that i need to squash before this gets merged

lrascao commented:
fixup rebase is done, commit history is now clean

cmeiklejohn commented:
Should this be merged now?  I don't remember where we left off with this and I'm trying to clear some of the outstanding PRs we have open.
